pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -46,35 +128,36 @@ func Endpoints(ds *v1.DiscoveryService, serviceClusters []string) *xdsapi.Discov
 		// responses. Pilot believes in eventual consistency and that at some point, Envoy
 		// will begin seeing results it deems to be good.
 		VersionInfo: version,
-		Nonce:       version,
+		Nonce:       version + "-" + time.Now().String(),
 	}
 
-	var totalEndpoints uint32
-	out.Resources = make([]types.Any, 0, len(serviceClusters))
-	for _, serviceCluster := range serviceClusters {
-		hostname, ports, labels := model.ParseServiceKey(serviceCluster)
-		instances, err := ds.Instances(hostname, ports.GetNames(), labels)
-		if err != nil {
-			log.Warnf("endpoints for service cluster %q returned error %q", serviceCluster, err)
-			continue
-		}
-		locEps := localityLbEndpointsFromInstances(instances)
-		if len(instances) == 0 {
-			log.Infoa("EDS: no instances ", serviceCluster, hostname, ports, labels)
-		}
-		clAssignment := &xdsapi.ClusterLoadAssignment{
-			ClusterName: serviceCluster,
-			Endpoints:   locEps,
+	out.Resources = make([]types.Any, 0, len(clusterNames))
+	for _, clusterName := range clusterNames {
+		clAssignmentRes := s.clusterEndpoints(ds, clusterName)
+		if clAssignmentRes != nil {
+			out.Resources = append(out.Resources, *clAssignmentRes)
 		}
-		//log.Infof("EDS: %v %s", serviceCluster, clAssignment.String())
-		clAssignmentRes, _ = types.MarshalAny(clAssignment)
-		out.Resources = append(out.Resources, *clAssignmentRes)
-		totalEndpoints += uint32(len(locEps))
 	}
 
 	return out
 }
 
+// Get the ClusterLoadAssignment for a cluster.
+func (s *DiscoveryServer) clusterEndpoints(ds *v1.DiscoveryService, clusterName string) *types.Any {
+	c := s.getOrAddEdsCluster(clusterName)
+	if c.LoadAssignment == nil { // fresh cluster
+		updateCluster(clusterName, c)
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
```

Comment:
nit: when a mutex guards only a single line IMO it's clearer to just surround the line with a lock/unlock (and lower runtime cost vs deferred func).

```go
c.mutex.Lock()
clAssignmentRes, _ := types.MarshalAny(c.LoadAssignment)
c.mutex.Unlock()
return clAssignmentRes
```

Reply:
Agree - not sure it's worth changing (before someone will say the opposite...).